### PR TITLE
Changes from systemapp spike

### PR DESCRIPTION
**Pull request by freddydk**
Beside the feednack from the systemapp spike, I also completed the scripts for LocalDevEnv and CloudDevEnv in this PR.

### CODE REVIEW

File: Actions/AL-Go-Helper.ps1 [hunk 1]
```diff
@@ -310,9 +312,10 @@ function ReadSettings {
         "installPerformanceToolkit"              = $false
         "enableCodeCop"                          = $false
         "enableUICop"                            = $false
+        "doNotBuildTests"                        = $false
         "doNotRunTests"                          = $false
         "appSourceCopMandatoryAffixes"           = @()
-        "memoryLimit"                            = "6G"
+        "memoryLimit"                            = ""
```
Comment (PooyaKharamesh) [inline Actions/AL-Go-Helper.ps1:318]: minor: IMO in almost all the cases memory usage is lower than 6GB but does that mean that we dont have any limits by default? what happens if a partner has a base app size app on github ?

File: Actions/AL-Go-Helper.ps1 [hunk 2]
```diff
@@ -815,7 +824,9 @@ function CreateDevEnv {
         [Parameter(Mandatory=$true, ParameterSetName='local')]
         [pscredential] $credential,
         [Parameter(ParameterSetName='local')]
-        [string] $containerName = ""
+        [string] $containerName = "",
+        [string] $insiderSasToken = "",
+        [string] $LicenseFileUrl = ""
```
Comment (PooyaKharamesh) [inline Actions/AL-Go-Helper.ps1:829]: does that mean that license should be available publicly as there is no authentication?

File: Actions/AL-Go-Helper.ps1 [hunk 3]
```diff
@@ -966,13 +974,18 @@ function CreateDevEnv {
...
                 $runAlPipelineParams += @{
+                    "updateLaunchJson" = "Local Sandbox ($containerName)"
                     "containerName" = $containerName
                 }
             }
+            else {
+                $runAlPipelineParams += @{
+                    "updateLaunchJson" = "Local Sandbox"
+                }
+            }
```
Comment (PooyaKharamesh) [inline Actions/AL-Go-Helper.ps1:988]: minor: code can be simplified

File: Actions/RunPipeline/RunPipeline.ps1 [hunk 4]
```diff
@@ -73,6 +77,7 @@ try {
     $artifact = $repo.artifact
     $installApps = $repo.installApps
     $installTestApps = $repo.installTestApps
+    $doNotBuildTests = $repo.doNotBuildTests
```
Comment (PooyaKharamesh) [inline Actions/RunPipeline/RunPipeline.ps1:80]: might be easier to read and avoiding double negative by consider using positive flags in this case BuildTests

File: Templates/AppSource App/.AL-Go/cloudDevEnv.ps1 [hunk 5]
```diff
@@ -1,25 +1,65 @@
...
+Write-Host @'
+This script will create a cloud based development environment (Business Central SaaS Sandbox) for your project.
+All apps and test apps will be compiled and published to the environment in the development scope.
+The script will also modify launch.json to have a "Cloud Sandbox (<name>)" configuration point to your environment.
+
+'@
+
+$settings = ReadSettings -baseFolder $baseFolder -userName $env:USERNAME
+
+Write-Host
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/cloudDevEnv.ps1:43]: does that do anything? I guess it is for adding an empty line ?

File: Templates/AppSource App/.AL-Go/cloudDevEnv.ps1 [hunk 6]
```diff
@@ -1,25 +1,65 @@
...
+        -default "$($env:USERNAME)-sandbox"
+}
 
-$reuseExistingEnvironment = Select-Value `
-    -title "What if the environment already exists?" `
-    -options @{ "Yes" = "Reuse existing environment"; "No" = "Recreate environment" } `
-    -question "Select behavior" `
-    -default "No"
+if (-not $PSBoundParameters.ContainsKey('reuseExistingEnvironment')) {
+    $reuseExistingEnvironment = (Select-Value `
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/cloudDevEnv.ps1:53]: could it be a mandatory parameter so we dont need to do this ?

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 7]
```diff
@@ -1,38 +1,109 @@
...
+'@
+
+Write-Host @'
+This script will create a docker based local development environment for your project.
+All apps and test apps will be compiled and published to the environment in the development scope.
+The script will also modify launch.json to have a Local Sandbox configuration point to your environment.
+
+'@
+
+$settings = ReadSettings -baseFolder $baseFolder -userName $env:USERNAME
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:44]: dont we have the settings in the env ?

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 8]
```diff
@@ -1,38 +1,109 @@
...
+
+Write-Host "Checking System Requirements"
+$dockerService = Get-Service docker -ErrorAction SilentlyContinue
+if (-not $dockerService -or $dockerService.Status -ne "Running") {
+    throw "Creating a local development enviroment requires you to have Docker installed and running."
+}
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:50]: minor:  we usually add a empty line after }

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 9]
```diff
@@ -1,38 +1,109 @@
...
+if ($settings.keyVaultName) {
+    if (-not (Get-Module -ListAvailable -Name 'Az.KeyVault')) {
+        throw "A keyvault name is defined in Settings, you need to have the Az.KeyVault PowerShell module installed or you can set the keyVaultName to an empty string in the user settings file ($($ENV:UserName).Settings.json)."
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:53]: can we add install-module command to  the message

Comment (PooyaKharamesh) [reply Templates/AppSource App/.AL-Go/localDevEnv.ps1:53]: if they have set a keyvault potentially they have secrets in the key vault. we can add that the secrets wont be read from keyvault anymore

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 10]
```diff
@@ -1,38 +1,109 @@
...
+    }
+}
+
+Write-Host
+
+if (-not $containerName) {
+    $containerName = Enter-Value `
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:60]: could this parameters be mandatory

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 11]
```diff
@@ -1,38 +1,109 @@
...
 }
-else {
-    $credential = Get-Credential -Message "Please enter username and password for your container" -UserName "admin"
+
+if (-not $licenseFileUrl) {
+    if ($settings.type -eq "AppSource App") {
+        $description = "When developing AppSource Apps, your local development environment needs the developer licensefile with permissions to your AppSource app object IDs"
+    }
+    else {
+        $description = "When developing PTEs, your can optionally specify a developer licensefile with permissions to object IDs of your dependant apps"
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:93]: you

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 [hunk 12]
```diff
@@ -1,38 +1,109 @@
...
+    }
+    $licenseFileUrl = Enter-Value `
```
Comment (PooyaKharamesh) [inline Templates/AppSource App/.AL-Go/localDevEnv.ps1:95]: shouldnt it pass for pte deployment? as it is optional

File: Templates/Per Tenant Extension/.AL-Go/cloudDevEnv.ps1 [hunk 13]
```diff
@@ -1,25 +1,65 @@
+#
+# Script for creating cloud development environment
+# Please do not modify this script as it will be auto-updated from the AL-Go Template
+# Recommended approach is to use as is or add a script (freddyk-devenv.ps1), which calls this script with the user specific parameters
+#
+Param(
```
Comment (PooyaKharamesh) [inline Templates/Per Tenant Extension/.AL-Go/cloudDevEnv.ps1:6]: code looks very similar to local dev env. can it be refactored ?

File: Templates/Per Tenant Extension/.AL-Go/localDevEnv.ps1 [hunk 14]
```diff
@@ -1,38 +1,109 @@
...
+}
+if ($settings.keyVaultName) {
+    if (-not (Get-Module -ListAvailable -Name 'Az.KeyVault')) {
+        throw "A keyvault name is defined in Settings, you need to have the Az.KeyVault PowerShell module installed or you can set the keyVaultName to an empty string in the user settings file ($($ENV:UserName).Settings.json)."
+    }
+}
+
+Write-Host
+
+if (-not $containerName) {
```
Comment (PooyaKharamesh) [inline Templates/Per Tenant Extension/.AL-Go/localDevEnv.ps1:59]: similar to the comments above

**Review by PooyaKharamesh — APPROVED** [review]

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 (hunk 12 above)
Comment (freddydk) [reply Templates/AppSource App/.AL-Go/localDevEnv.ps1:95]: Good catch.
I have added a default for PTEs which results in no licensefile.
If you enter "none" as license file appsource apps you also get no license file - not sure anybody will use that, but the option is there.

File: Actions/AL-Go-Helper.ps1 (hunk 3 above)
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:988]: I don't think so - it is important that the ContainerName is NOT specified in the params if the value is blank - and if I want the ( ) around the containername in local sandbox, I think I need the if.
If this is not what you had in mind, let me know.

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 (hunk 9 above)
Comment (freddydk) [reply Templates/AppSource App/.AL-Go/localDevEnv.ps1:53]: this script is only running locally - there we cannot reach the github secrets - only the keyvault.

File: Templates/AppSource App/.AL-Go/localDevEnv.ps1 (hunk 7 above)
Comment (freddydk) [reply Templates/AppSource App/.AL-Go/localDevEnv.ps1:44]: Not when running locally

File: Actions/RunPipeline/RunPipeline.ps1 (hunk 4 above)
Comment (freddydk) [reply Actions/RunPipeline/RunPipeline.ps1:80]: Could be - will think about that one for a refactor task

**Pull request merged**